### PR TITLE
API Purge

### LINKED ISSUE #401: Invalid BBox Causes 500 Error
I would expect an error message with a 400 status code.

```
 ~ $ http 'https://dev.plenar.io/api/v2/data-sets/chicago-beach-lab-data-culture-tests?bbox={"type":"Polygon", "coordinates":[[999,999],[999,-999],[-999,-999],[-999,999],[999,999]]}'
HTTP/1.1 500 Internal Server Error
Connection: keep-alive
Content-Length: 20
Date: Mon, 06 Aug 2018 15:41:43 GMT
Server: nginx/1.10.3 (Ubuntu)
access-control-allow-credentials: true
access-control-allow-origin: *
access-control-expose-headers: 
cache-control: max-age=0, private, must-revalidate
x-request-id: 2l45boijbijvqneuso000rv1

Something went wrong
```

### LINKED ISSUE #403: API List Endpoint list Too Much Info
We really only want public facing information here. There's just too much being dumped that we really don't want bleeding out.

## What happened?
```json
{
  "attribution": "City of Chicago",
  "bbox": {
    "coordinates": [
      [
        [
          42.8435743,
          -118.492032
        ],
        [
          34.010177,
          -118.492032
        ],
        [
          34.010177,
          -87.0158197949677
        ],
        [
          42.8435743,
          -87.0158197949677
        ],
        [
          42.8435743,
          -118.492032
        ]
      ]
    ],
    "srid": 4326
  },
  "description": null,
  "first_import": "2018-08-02T16:25:00.006941Z",
  "id": 66,
  "inserted_at": "2018-07-17T17:45:49.060174Z",
  "latest_import": "2018-08-06T16:26:00.524113Z",
  "name": "Chicago Area Alternative Fuel Locations",
  "next_import": "2018-08-07T16:25:00.043483Z",
  "refresh_ends_on": null,
  "refresh_interval": 1,
  "refresh_rate": "days",
  "refresh_starts_on": "2018-01-01",
  "slug": "chicago-area-alternative-fuel-locations",
  "source_type": "csv",
  "source_url": "https://data.cityofchicago.org/api/views/f7f2-ggz5/rows.csv?accessType=DOWNLOAD",
  "state": "ready",
  "table_name": "chicago_alternative_fuel_locations_chicago",
  "time_range": {
    "lower": "1995-01-15T00:00:00",
    "lower_inclusive": true,
    "upper": "2018-08-06T04:04:59",
    "upper_inclusive": true
  },
  "updated_at": "2018-08-06T16:26:00.552286Z",
  "user_id": 1
}
```

## What did you expect?
```json
{
  "attribution": "City of Chicago",
  "bbox": {
    "coordinates": [
      [
        [
          42.8435743,
          -118.492032
        ],
        [
          34.010177,
          -118.492032
        ],
        [
          34.010177,
          -87.0158197949677
        ],
        [
          42.8435743,
          -87.0158197949677
        ],
        [
          42.8435743,
          -118.492032
        ]
      ]
    ],
    "srid": 4326
  },
  "description": null,
  "first_import": "2018-08-02T16:25:00.006941Z",
  "latest_import": "2018-08-06T16:26:00.524113Z",
  "name": "Chicago Area Alternative Fuel Locations",
  "next_import": "2018-08-07T16:25:00.043483Z",
  "refresh_ends_on": null,
  "refresh_interval": 1,
  "refresh_rate": "days",
  "refresh_starts_on": "2018-01-01",
  "slug": "chicago-area-alternative-fuel-locations",
  "source_type": "csv",
  "source_url": "https://data.cityofchicago.org/api/views/f7f2-ggz5/rows.csv?accessType=DOWNLOAD",
  "time_range": {
    "lower": "1995-01-15T00:00:00",
    "lower_inclusive": true,
    "upper": "2018-08-06T04:04:59",
    "upper_inclusive": true
  },
  "user": { "user": "object" }
}
```

This applies to `:get`, `:head` and `:describe`.

### LINKED ISSUE #404: Detail Endpoint doesn't have the right keys
```
     left:  ["Beach", "Culture Note", "Culture Reading Mean", "Culture Sample 1 Reading", "Culture Sample 1 Timestamp", "Culture Sample 2 Reading", "Culture Sample 2 Timestamp", "Culture Sample Interval", "Culture Test ID", "DNA Reading Mean", "DNA Sample 1 Reading", "DNA Sample 2 Reading", "DNA Sample Timestamp", "DNA Test ID", "Latitude", "Location", "Longitude", "row_id", "vpf_HEPt_EqddTtnO4Kq"]
     right: ["row_id", "DNA Test ID", "DNA Sample Timestamp", "Beach", "DNA Sample 1 Reading", "DNA Sample 2 Reading", "DNA Reading Mean", "Culture Test ID", "Culture Sample 1 Timestamp", "Culture Sample 1 Reading", "Culture Sample 2 Reading", "Culture Reading Mean", "Culture Note", "Culture Sample Interval", "Culture Sample 2 Timestamp", "Latitude", "Longitude", "Location" | "vpf_HEPt_EqddTtnO4Kq"]
```

### LINKED ISSUE #405: API Detail Endpoint 500 When Given ID Instead of Slug
```
  2) test GET detail endpoint will 404 when given the id, even if it's for a valid, ready data set (PlenarioWeb.Api.DetailControllerTest)
     test/plenario_web/controllers/api/detail_controller_test.exs:114
     ** (exit) exited in: GenServer.call(Plenario.ModelRegistry, {:lookup, "7220"}, 5000)
         ** (EXIT) an exception was raised:
             ** (KeyError) key "7220" not found in: %{7220 => :"Model.chicagobeachlabdnatests", "chicago-beach-lab-dna-tests" => :"Model.chicagobeachlabdnatests"}
                 (stdlib) :maps.get("7220", %{7220 => :"Model.chicagobeachlabdnatests", "chicago-beach-lab-dna-tests" => :"Model.chicagobeachlabdnatests"})
                 (plenario) lib/plenario/model_registry.ex:97: Plenario.ModelRegistry.handle_call/3
                 (stdlib) gen_server.erl:636: :gen_server.try_handle_call/4
                 (stdlib) gen_server.erl:665: :gen_server.handle_msg/6
                 (stdlib) proc_lib.erl:247: :proc_lib.init_p_do_apply/3
     code: |> get(detail_path(conn, :get, meta.id))
     stacktrace:
       (elixir) lib/gen_server.ex:831: GenServer.call/3
       (plenario) lib/plenario_web/controllers/api/detail_controller.ex:96: PlenarioWeb.Api.DetailController.construct_query_from_conn_assigns/2
       (plenario) lib/plenario_web/controllers/api/detail_controller.ex:109: PlenarioWeb.Api.DetailController.get/2
       (plenario) lib/plenario_web/controllers/api/detail_controller.ex:1: PlenarioWeb.Api.DetailController.action/2
       (plenario) lib/plenario_web/controllers/api/detail_controller.ex:1: PlenarioWeb.Api.DetailController.phoenix_controller_pipeline/2
       (plenario) lib/plenario_web/endpoint.ex:1: PlenarioWeb.Endpoint.instrument/4
       (phoenix) lib/phoenix/router.ex:278: Phoenix.Router.__call__/1
       (plenario) lib/plug/error_handler.ex:64: PlenarioWeb.Router.call/2
       (plenario) lib/plenario_web/endpoint.ex:1: PlenarioWeb.Endpoint.plug_builder_call/2
       (plenario) lib/plenario_web/endpoint.ex:1: PlenarioWeb.Endpoint.call/2
       (phoenix) lib/phoenix/test/conn_test.ex:224: Phoenix.ConnTest.dispatch/5
       test/plenario_web/controllers/api/detail_controller_test.exs:119: (test)
```

### LINKED ISSUE #406: API Detail @describe Endpoint Doesn't Give the Right Info
```
  4) test GET @describe endpoint the response object exactly contains meta information and field information (PlenarioWeb.Api.DetailControllerTest)
     test/plenario_web/controllers/api/detail_controller_test.exs:211
     Assertion with == failed
     code:  assert Map.keys(x1) == @describe_keys
     left:  ["Beach", "Culture Note", "Culture Reading Mean", "Culture Sample 1 Reading", "Culture Sample 1 Timestamp", "Culture Sample 2 Reading", "Culture Sample 2 Timestamp", "Culture Sample Interval", "Culture Test ID", "DNA Reading Mean", "DNA Sample 1 Reading", "DNA Sample 2 Reading", "DNA Sample Timestamp", "DNA Test ID", "Latitude", "Location", "Longitude", "row_id", "vpf_AC5CQWUSZ0IO5nZf"]
     right: ["attribution", "bbox", "description", "first_import", "latest_import", "name", "next_import", "refresh_ends_on", "refresh_interval", "refresh_rate", "refresh_starts_on", "slug", "source_url", "time_range", "user", "fields", "virtual_date_fields", "virtual_point_fields"]
     stacktrace:
       (elixir) lib/enum.ex:737: Enum."-each/2-lists^foreach/1-0-"/2
       (elixir) lib/enum.ex:737: Enum.each/2
       test/plenario_web/controllers/api/detail_controller_test.exs:221: (test)
```

### LINKED ISSUE #407: API List Endpoint Doesn't Filter on Time Range
```
  1) test filter list endpoint by time range with a well formatted time range (PlenarioWeb.Api.ListControllerTest)
     test/plenario_web/controllers/api/list_controller_test.exs:209
     ** (Ecto.SubQueryError) the following exception happened when compiling a subquery.
     
         ** (Ecto.Query.CastError) lib/plenario_web/controllers/api/utils.ex:140: value `~N[2016-05-26 00:00:00]` in `where` cannot be cast to type Plenario.TsRange in query:
         
         from m in Plenario.Schemas.Meta,
           where: m.state == "ready",
           where: m.time_range >= ^~N[2016-05-26 00:00:00],
           where: m.time_range <= ^~N[2017-05-26 00:00:00]
         
     
     The subquery originated from the following query:
     
     from m in subquery(from m in Plenario.Schemas.Meta,
       where: m.state == "ready",
       where: m.time_range >= ^~N[2016-05-26 00:00:00],
       where: m.time_range <= ^~N[2017-05-26 00:00:00]),
       select: count("*")
     
     code: |> get(list_path(conn, :get, %{time_range: "in:#{@good_time_range}"}))
     stacktrace:
       (ecto) lib/ecto/repo/queryable.ex:124: Ecto.Repo.Queryable.execute/5
       (ecto) lib/ecto/repo/queryable.ex:37: Ecto.Repo.Queryable.all/4
       (ecto) lib/ecto/repo/queryable.ex:70: Ecto.Repo.Queryable.one/4
       (scrivener_ecto) lib/scrivener/paginater/ecto/query.ex:37: Scrivener.Paginater.Ecto.Query.total_entries/3
       (scrivener_ecto) lib/scrivener/paginater/ecto/query.ex:10: Scrivener.Paginater.Ecto.Query.paginate/2
       (plenario) lib/plenario_web/controllers/api/list_controller.ex:102: PlenarioWeb.Api.ListController.get/2
       (plenario) lib/plenario_web/controllers/api/list_controller.ex:1: PlenarioWeb.Api.ListController.action/2
       (plenario) lib/plenario_web/controllers/api/list_controller.ex:1: PlenarioWeb.Api.ListController.phoenix_controller_pipeline/2
       (plenario) lib/plenario_web/endpoint.ex:1: PlenarioWeb.Endpoint.instrument/4
       (phoenix) lib/phoenix/router.ex:278: Phoenix.Router.__call__/1
       (plenario) lib/plug/error_handler.ex:64: PlenarioWeb.Router.call/2
       (plenario) lib/plenario_web/endpoint.ex:1: PlenarioWeb.Endpoint.plug_builder_call/2
       (plenario) lib/plenario_web/endpoint.ex:1: PlenarioWeb.Endpoint.call/2
       (phoenix) lib/phoenix/test/conn_test.ex:224: Phoenix.ConnTest.dispatch/5
       test/plenario_web/controllers/api/list_controller_test.exs:212: (test)



  2) test filter list endpoint by time range will 400 with a poorly formatted time range (PlenarioWeb.Api.ListControllerTest)
     test/plenario_web/controllers/api/list_controller_test.exs:225
     ** (Ecto.SubQueryError) the following exception happened when compiling a subquery.
     
         ** (Ecto.Query.CastError) lib/plenario_web/controllers/api/utils.ex:190: value `{"in", "whenever"}` in `where` cannot be cast to type Plenario.TsRange in query:
         
         from m in Plenario.Schemas.Meta,
           where: m.state == "ready",
           where: m.time_range == ^{"in", "whenever"}
         
     
     The subquery originated from the following query:
     
     from m in subquery(from m in Plenario.Schemas.Meta,
       where: m.state == "ready",
       where: m.time_range == ^{"in", "whenever"}),
       select: count("*")
     
     code: |> get(list_path(conn, :get, %{time_range: "in:#{@bad_time_range}"}))
     stacktrace:
       (ecto) lib/ecto/repo/queryable.ex:124: Ecto.Repo.Queryable.execute/5
       (ecto) lib/ecto/repo/queryable.ex:37: Ecto.Repo.Queryable.all/4
       (ecto) lib/ecto/repo/queryable.ex:70: Ecto.Repo.Queryable.one/4
       (scrivener_ecto) lib/scrivener/paginater/ecto/query.ex:37: Scrivener.Paginater.Ecto.Query.total_entries/3
       (scrivener_ecto) lib/scrivener/paginater/ecto/query.ex:10: Scrivener.Paginater.Ecto.Query.paginate/2
       (plenario) lib/plenario_web/controllers/api/list_controller.ex:102: PlenarioWeb.Api.ListController.get/2
       (plenario) lib/plenario_web/controllers/api/list_controller.ex:1: PlenarioWeb.Api.ListController.action/2
       (plenario) lib/plenario_web/controllers/api/list_controller.ex:1: PlenarioWeb.Api.ListController.phoenix_controller_pipeline/2
       (plenario) lib/plenario_web/endpoint.ex:1: PlenarioWeb.Endpoint.instrument/4
       (phoenix) lib/phoenix/router.ex:278: Phoenix.Router.__call__/1
       (plenario) lib/plug/error_handler.ex:64: PlenarioWeb.Router.call/2
       (plenario) lib/plenario_web/endpoint.ex:1: PlenarioWeb.Endpoint.plug_builder_call/2
       (plenario) lib/plenario_web/endpoint.ex:1: PlenarioWeb.Endpoint.call/2
       (phoenix) lib/phoenix/test/conn_test.ex:224: Phoenix.ConnTest.dispatch/5
       test/plenario_web/controllers/api/list_controller_test.exs:227: (test)
```

### LINKED ISSUE #408: API Detail Endpoint Doesn't Filter on Time Range
```
  1) test filter results by time range will 400 with a poorly formatted time range (PlenarioWeb.Api.DetailControllerTest)
     test/plenario_web/controllers/api/detail_controller_test.exs:243
     ** (RuntimeError) expected response with status 400, got: 200, with body:
    ... got a 200

  2) test filter results by time range with a well formatted time range (PlenarioWeb.Api.DetailControllerTest)
     test/plenario_web/controllers/api/detail_controller_test.exs:227
     Assertion with == failed
     code:  assert length(res["data"]) == 1
     left:  500
     right: 1
     stacktrace:
       test/plenario_web/controllers/api/detail_controller_test.exs:233: (test)
```

### LINKED ISSUE #415: Shim API: Datasets Endpoint Doesn't Conform to V1 Spec
Title says it all.

### LINKED ISSUE #416: Shim API: Detail Endpoint Doesn't Conform to V1 Spec

### LINKED ISSUE #417: Shim API: Missing Fields Endpoint

### LINKED ISSUE #418: MetaActions :compute_bbox! Flips X and Y
When creating any geometry, the correct order of coords is `x, y`, not `y, x`. This one's on me.

### PR DESCRIPTION
**Pull request by vforgione**
So, I kinda sorta totally gutted and rewrote the API. I have a good
reason for doing this, I swear.

We had a handful of seemingly unrelated bugs in the API, and being the
good Orkin Man that I am I started fixing them. At first I was doing
them one at a time, but then that wasn't really fixing anything -- it
was more like handling edge cases. So then I started looking for
patterns in the bugs, and how far down the stack they could go.

The way I started this was with the tests. They sucked. I'm sorry for
being so blunt, but they really weren't that great. I probably should
have scrutinized them earlier when the PRs came up, but I think that was
back in late April/early May when I was getting married and buying a
house. Not an excuse, just some context. Or at least that's when the
bulk of the API was being developed.

Anywho, I rewrote the tests and, lo and behold, it turns out that we
were not delivering on what we promised. We also had a ton of bugs, and
some really sloppy hacks to force things together.

At the same time, I asked the CDO folks to start hammering the Shim API
and they found tons of broken stuff: the datasets endpoint didn't return
the same shape as the existing app, the fields endpoint was missing,
they didn't even try the detail endpoint but my inspection found it to
also be mishapened.

Given all of that, it seemed like giving this a fresh start was the best
approach. The controllers are a lot slimmer, the views delegate where
appropriate, the plugs are now purely functions, the utils are cleanly
laid out, and best of all the errors are all being caught and handled
properly.

I've also done away with all references to 422 status codes. I'm sorry,
that's just crazy talk. 400 is the right way to do it -- it's a bad
request.

Fixes #401
Fixes #403
Fixes #404
Fixes #405
Fixes #406
Fixes #407
Fixes #408
Fixes #415
Fixes #416
Fixes #417
Fixes #418

### CODE REVIEW

File: test/plenario_web/controllers/api/list_controller_test.exs [hunk 1]
```diff
@@ -3,7 +3,14 @@ defmodule PlenarioWeb.Api.ListControllerTest do
```
Comment (vforgione) [inline test/plenario_web/controllers/api/list_controller_test.exs]: This was also a total rewrite. Look at the file, not the diff.

File: test/plenario_web/controllers/api/detail_controller_test.exs [hunk 2]
```diff
@@ -3,7 +3,14 @@ defmodule PlenarioWeb.Api.DetailControllerTest do
```
Comment (vforgione) [inline test/plenario_web/controllers/api/detail_controller_test.exs]: This was also a total rewrite. Look at the file, not the diff.

File: test/plenario_web/controllers/api/shim_controller_test.exs [hunk 3]
```diff
@@ -3,6 +3,10 @@ defmodule PlenarioWeb.Api.ShimControllerTest do
```
Comment (vforgione) [inline test/plenario_web/controllers/api/shim_controller_test.exs]: This was also a total rewrite. Look at the file, not the diff.

**Comment by vforgione** [comment]
Also fixes #411 and fixes #412

**Comment by brlodi** [comment]
I'm willing to have the fight over status codes (though I agree in most cases 422 is wrong).

I believe they should be applied as follows:

**Page**
- `page=wrong` – **400 Bad Request**
- `page=0` – **200 OK** I would generally expect an API to ignore falsy values for non-boolean parameters or take them as an explicit direction to "use defaults"; I would not make this an HTTP error at all, and simply add a note to the documentation
- `page=-3` – **400 Bad Request** (though it'd be great to offer reverse indexing like Python, so e.g. -1 selects the last page, in which case it would be valid)
- `page=999999999999` – Probably should just return the very last (necessary incomplete) page, but if we must error then it's absolutely a **422**; RFC 4918 states that **422 Unprocessable Entity** is applicable when the request entity is syntactically well-formed, but semantically erroneous.

**Page Size**
- `page_size=wrong` – **400 Bad Request**
- `page_size=0` – **200 OK** It seems straightforward enough to return the default size of page; see arguments for `page=0` above
- `page_size=-232` – **400 Bad Request**
- `page_size=1276971623108` – **403 Forbidden** because we absolutely _do_ understand the request, but we just refuse to do it because it would be a huge response

As a summary statement, I'd like to reiterate that HTTP Status Codes are intended to describe the status of the HTTP transaction itself. It is perfectly reasonable to return some amount of error status under a **200 Accepted** code if the error is unrelated to the request itself, e.g. "give me the first 100 records of dataset `foobar`" could very well merit a response of `200 OK {"error": "Foo Bar has not had its first ingest yet, so no records are available"}` (I realize that doesn't match our return spec, but the point is the request is 100% valid but there is a business behavior issue with responding)

**Comment by vforgione** [comment]
Page should be a _positive, non zero_ integer; otherwise, it's a bad request. This goes the same for page size.

I'm not entirely convinced that `0` is a valid value, especially if we're strictly evaluating and enforcing these values.

403 seems inappropriate -- I would expect to see that for a resource I cannot access, not a query that I tried to ham fist.

422 means that we've fully interpreted the value and determined that we don't like it. The plugs only check for type and lower bound range validity and pass the request along. If a user passes `page=999999999` then it's up to the database to determine if that combination of `limit/offset` is an error.

I've also never seen 422 used -- I know it's a legit status per the RFCs, but I have yet to see it in the wild. I think most clients are smart enough to determine that a 4xx status is an error, but are any of them actually going to handle the robust error message and adjust accordingly?

Additionally, I just hate it. I don't know why. I have this really awful, negative emotional reaction to it :)

**Comment by vforgione** [comment]
While working on the docs, I found some funky stuff I don't really like in the AoT endpoint. Hold up on reviewing that for now.

**Comment by vforgione** [comment]
Ok, fixed the AoT funk.

File: lib/plenario_web/controllers/api/utils.ex [hunk 4]
```diff
@@ -254,4 +294,50 @@ defmodule PlenarioWeb.Api.Utils do
...
+
+  def apply_filter(query, fname, "le", value), do: where(query, [q], field(q, ^fname) <= ^value)
+
+  def apply_filter(query, fname, "eq", value), do: where(query, [q], field(q, ^fname) == ^value)
+
+  def apply_filter(query, fname, "ge", value), do: where(query, [q], field(q, ^fname) >= ^value)
+
+  def apply_filter(query, fname, "gt", value), do: where(query, [q], field(q, ^fname) > ^value)
+
+  def apply_filter(query, fname, "in", value), do: where(query, [q], field(q, ^fname) in ^value)
```
Comment (HeyZoos) [inline lib/plenario_web/controllers/api/utils.ex]: These filters are just renaming the `where_condition` functions but you've completely deleted all the comments. Why? This shoots us in the foot in the future.

File: lib/plenario_web/controllers/api/aot_controller.ex [hunk 5]
```diff
@@ -1,236 +1,108 @@
...
-    render_page(conn, "get.json", conn.params, page.entries, page, true)
+    try do
+      page = conn.assigns[:page]
+      page_size = conn.assigns[:page_size]
+      data = Repo.paginate(query, page: page, page_size: page_size)
+      {:ok, data}
+    rescue
+      e in [Ecto.QueryError, Ecto.SubQueryError, Postgrex.Error] ->
+        {:error, e.message}
+    end
```
Comment (HeyZoos) [inline lib/plenario_web/controllers/api/aot_controller.ex]: Handling errors like this makes it tough to distinguish between what is bad user input and what is a true error on our part. At the very least, `halt_with` should send a copy of the error to sentry, or this head ` defp render_data({:error, message}, conn, _)` of `render_data` should send information to sentry. Sweeping `try-rescues` also runs counter to the Erlang "let it die" philosophy. Bad input shouldn't cause us to crash, bad input we handle incorrectly should.

**Comment by HeyZoos** [comment]
I'd also like to :+1: Ben's comment about reverse indexing. It would be a nice touch.

File: lib/plenario_web/controllers/api/aot_controller.ex (hunk 5 above)
Comment (vforgione) [reply lib/plenario_web/controllers/api/aot_controller.ex]: The reason for this is because I got rid of the explicit field naming in the plugs -- if a user passes a param `foo=bar` and there isn't a `foo` field on the data set, it will error.

This isn't an uncaught internal error on our part, this is a client error. I catch it and send it back to them.
